Rwjs/session fix m11 - Fix for making sure that sessions are removed from session_map

robsimmonds opened the pull request.

robsimmonds committed:
Added a static SessionManager share_ptr to the OnMessageTask class so that
a pointer to an end-of-life session can be removed from the
SessionManager session map.

robsimmonds committed:
Set the pointer to the SessionManager for the OnMessageTask objects.

robsimmonds committed:
Updated some logging messages.

Modefined the way that sessions get deleted to make sure they all get
removed from the SessionManager session map.

robsimmonds committed:
Removed a test in the excution loop that was not needed.

robsimmonds committed:
Updated some logging.

Made change to session deletion so that it is consistent in both cases when
a WebSocket closes and when the last OnMessageTask referencing a session
is deleted.

robsimmonds committed:
Added shared_ptr to SessionManager to enable sessions to be removed
from the SessionManager session map.

robsimmonds committed:
Made _num_sessions counter volatile.

confluence commented:
What's the easiest way to reproduce this bug for testing purposes?

The CI unit tests are all crashing (with a segfault in `MomentTest.CheckConsistency`). I will check if I can reproduce this locally.

confluence commented on src/Session/OnMessageTask.h:
```diff
@@ -31,10 +35,14 @@ class OnMessageTask {
         }
         _session = nullptr;
     }
+    static void SetSessionManager(shared_ptr<SessionManager>& sm) {
```
I suggest renaming the parameter to `session_manager`.

confluence commented:
I get a segfault in `IcdTest.RegionRegister` when I run the tests locally.

robsimmonds commented:
This problem is likely due to the SessionManager pointer not being set in the testing backend. It looked to me that only some of the tests used the SessionManager so either one will have to be created for the testing backend, or an extra test will have to be put in the code to prevent that shared_ptr having the method called on it in ~OnMessageTask. When I tried the tests locally they seemed to be working, even though I was not expecting them to. If I can get the tests to fail I would likely be able to fix the problem.

robsimmonds committed:
Merge branch 'dev' into rwjs/session_fix_m11

robsimmonds committed:
Added check before using the shared_ptr to the SessionManager. Turns
out that it behaves differently depending on what compiler settings
are used with g++ v9.4.0.

robsimmonds committed:
Changed from a short to a long parameter name.

ajm-ska commented:
@robsimmonds I know the Moment ICD tests on Ubuntu 20.04 and 21.10 appear to continuously fail on this branch and the `rwjs/session_fix_m11` branch for some reason. It may not be due to any changes in these branches because I did notice an occasional failure of the Moment tests occurring in other branches recently. I'm just surprised that they are consistently failing here for some reason, plus only failing on 2 out of 7 systems. We (@acdo2002!) can investigate more on Monday. @acdc2002 recently reported that the Ubuntu contour tests on the 'dev' branch take 10% longer than before. I wonder if whatever caused that is having the same effect here. Maybe we just need to increase a timeout on the Moment ICD tests to make them less strict.

robsimmonds commented:
> @robsimmonds I know the Moment ICD tests on Ubuntu 20.04 and 21.10 appear to continuously fail on this branch and the `rwjs/session_fix_m11` branch for some reason. It may not be due to any changes in these branches because I did notice an occasional failure of the Moment tests occurring in other branches recently. I'm just surprised that they are consistently failing here for some reason, plus only failing on 2 out of 7 systems. We (@acdo2002!) can investigate more on Monday. @acdc2002 recently reported that the Ubuntu contour tests on the 'dev' branch take 10% longer than before. I wonder if whatever caused that is having the same effect here. Maybe we just need to increase a timeout on the Moment ICD tests to make them less strict.

OK, thanks it is strange. Is there a good way for me to download the complete set of test files so I can also see if they run on my ubuntu test system. I know they used to be in google, which I have problems downloading to from our research cluster, but could download them to my home machine.

ajm-ska commented:
@robsimmonds Unfortunately we don't have a definitive list of the image files currently used by the ICD tests. We read from a folder called `set_QA`. It is ~270GB in size so it is not practical to share it. But I'm fairly sure that not all of the images are actually used, so the folder could be significantly smaller. 

I will create a new issue on the ICD-test repository for us to make a list of the files that are actually used in the tests as it would be useful to know.

For now I just checked the Moment tests. I put the images they use in a folder, and temporarily uploaded it to our ASIAA webserver (we can find a better location later). It is 28GB (20GB zipped)! It unzips to `set_QA`. 
You can download it here:
`wget https://alma.asiaa.sinica.edu.tw/_downloads/ICD_Moments_images.tgz`

I tried running the Moment tests on an Ubuntu 20.04 server natively and the tests pass fine using your branch. So maybe the Ubuntu Docker containers don't have enough resources for some reason, even though they have been working fine before.

If you would like to try it yourself, here is a recap on how to run the ICD tests:
1. Prepare the ICD tests
```
git clone https://github.com/CARTAvis/carta-backend-ICD-test.git
cd carta-backend-ICD-test
git submodule init
git submodule update
npm install
cd protobuf
./build_proto.sh
cd ..
```
Modify `src/test/config.json` to set the correct serverURL. e.g
`"serverURL": "ws://127.0.0.1:3002",`

2. Start the carta-backend pointing towards the parent folder that contains the `set_QA` folder. e.g 
`./carta_backend /images --port 3002 --debug_no_auth --no_frontend`

3. In the `carta-backend-ICD-test` folder, run the Moment ICD tests:
```
CI=true npm test src/test/MOMENTS_GENERATOR_CASA.test.ts
CI=true npm test src/test/MOMENTS_GENERATOR_EXCEPTION.test.ts
CI=true npm test src/test/MOMENTS_GENERATOR_FITS.test.ts
CI=true npm test src/test/MOMENTS_GENERATOR_HDF5.test.ts
CI=true npm test src/test/MOMENTS_GENERATOR_SAVE.test.ts
CI=true npm test src/test/MOMENTS_GENERATOR_CANCEL.test.ts
```

robsimmonds commented:
@ajm-asiaa thanks for looking into this. It would be good if all the ICD test files were available for download, but that might be difficult given their size. It might be good to mention on the ICD tests github page that the ICD tests need a set of files that are not currently in a public repository just so people don't try to run the tests and think they are broken.

robsimmonds committed:
Shorter timeout on exit.

robsimmonds committed:
Merge branch 'dev' into rwjs/session_fix_m11

robsimmonds committed:
Added _dummy_backend->WaitForJobFinished() to RegionRegister test.
